packages/bull-vault/src/AuctionBull.sol

```diff
@@ -0,0 +1,565 @@
+        } else {
+            IBullStrategy(bullStrategy).redeemCrabAndWithdrawWEth(_crabAmount, wPowerPerpAmount);
+
+            {
+                uint256 remainingAmount = wPowerPerpAmount;
+                uint256 ordersLength = _orders.length;
+                for (uint256 i; i < ordersLength; ++i) {
+                    _transferToOrder(_orders[i], remainingAmount, _clearingPrice);
+                    if (remainingAmount > _orders[i].quantity) {
+                        remainingAmount = remainingAmount.sub(_orders[i].quantity);
+                    } else {
+                        break;
+                    }
+                }
+            }
+
+            _rebalanceLeverageComponent(_wethTargetInEuler, _wethLimitPrice);
+        }
+    }
+
+    function _executeCrabDeposit(uint256 _crabAmount, uint256 _wethTargetInEuler, uint256 _wethLimitPrice, uint256 _ethInCrab) internal {
+        uint256 wethFromEuler;
+        uint256 ethNeededForCrab = _crabAmount.wdiv(IERC20(crab).totalSupply()).wmul(_ethInCrab);
+        uint256 wethInCollateral = IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy));
+        if (_wethTargetInEuler > wethInCollateral) {
+            uint256 wethToBuy;
+            uint256 usdcToBorrow;
+            {
+                uint256 ethUsdPrice = UniOracle._getTwap(ethUSDCPool, weth, usdc, TWAP, false);
+                wethToBuy = _wethTargetInEuler.sub(wethInCollateral);
+                usdcToBorrow = wethToBuy.wmul(ethUsdPrice).div(1e12);
+            }
+            _exactOutFlashSwap(
+                usdc,
+                weth,
+                3000,
+                wethToBuy,
+                usdcToBorrow.wmul(_wethLimitPrice),
+                uint8(FLASH_SOURCE.FULL_REBALANCE_BORROW_USDC_BUY_WETH),
+                abi.encodePacked(wethToBuy, usdcToBorrow)
+            );
+        } else {
+            wethFromEuler = wethInCollateral.sub(_wethTargetInEuler);
+            IBullStrategy(bullStrategy).repayAndWithdrawFromLeverage(0, wethFromEuler);
+        }
+        ethNeededForCrab = ethNeededForCrab.sub(wethFromEuler);
+        IBullStrategy(bullStrategy).depositEthIntoCrab(ethNeededForCrab);
+    }
+
+    function _rebalanceLeverageComponent(uint256 _wethTargetInEuler, uint256 _wethLimitPrice) internal {
+        uint256 remainingWeth = IERC20(weth).balanceOf(address(this));
+        uint256 wethInCollateral = IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy));
+        if (_wethTargetInEuler > remainingWeth.add(wethInCollateral)) {
+            uint256 ethUsdPrice = UniOracle._getTwap(ethUSDCPool, weth, usdc, TWAP, false);
+            uint256 wethToBuy = _wethTargetInEuler.sub(remainingWeth.add(wethInCollateral));
+            uint256 usdcToBorrow = wethToBuy.wmul(ethUsdPrice).div(1e12);
+            _exactOutFlashSwap(
+                usdc,
+                weth,
+                3000,
+                wethToBuy,
+                usdcToBorrow.wmul(_wethLimitPrice),
```

Comment:
should use the wethToBuy.wmul(_wethLimitPrice) and then adjust for decimals (ie wmul(1e12))

Reply:
same comment as for the deposit crab flow trade